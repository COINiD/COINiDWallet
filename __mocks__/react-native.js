import MockAsyncStorage from 'mock-async-storage';

import { WebSocket } from 'mock-socket';

const rn = require('react-native');

global.fetch = require('jest-fetch-mock');

global.fetch.mockResponse(JSON.stringify({ data: '12345' }));

global.WebSocket = WebSocket;

jest.useFakeTimers();

jest.mock('Alert', () => ({ alert: jest.fn() }));

jest.mock('AsyncStorage', () => new MockAsyncStorage());

jest.mock('NativeAnimatedHelper');

jest.mock('NativeModules', () => ({
  P2PTransferBLECentralModule: {
    addListener: jest.fn(),
  },
  P2PTransferBLEPeripheralModule: {},
  KeyboardObserver: {
    addListener: jest.fn(),
  },
  StatusBarManager: {
    HEIGHT: 42,
    setColor: jest.fn(),
    setStyle: jest.fn(),
    setHidden: jest.fn(),
    setNetworkActivityIndicatorVisible: jest.fn(),
    setBackgroundColor: jest.fn(),
    setTranslucent: jest.fn(),
  },
}));

module.exports = rn;
