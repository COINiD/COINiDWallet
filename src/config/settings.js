export default {
  coin : 'testnet',
  appReturnScheme : 'coinid-tbtc',
  coldWalletMode: true,
  language: 'system',
  availableLanguages : ['system','en','de','fr','es','pt','eo','sv'],
  currency: 'USD',
  availableCurrencies : ['USD','EUR','JPY','GBP','CAD','CHF','KRW','NZD','SEK','NOK','DKK'],
  ranges: ['day', 'week', 'month', 'year'],
  refreshRates: {
    exchangeRate: 600000, // 10 minutes
    account: 600000, // 10 minutes
  },
  coldTransportTypes: [
    {
      key: '',
      title: 'settings.coldtransport.showdialog.title',
      description: 'settings.coldtransport.showdialog.description',
    },
    {
      key: 'qr',
      title: 'settings.coldtransport.qrcodes.title',
      description: 'settings.coldtransport.qrcodes.description',
    },
    {
      key: 'ble',
      title: 'settings.coldtransport.ble.title',
      description: 'settings.coldtransport.ble.description',
    },
  ],
  lockDurations: [
    { milliseconds: 0, title: 'settings.lockduration.immediately' },
    { milliseconds: 60000, title: 'settings.lockduration.1minute' },
    { milliseconds: 300000, title: 'settings.lockduration.5minutes' },
    { milliseconds: 900000, title: 'settings.lockduration.15minutes' },
    { milliseconds: 1800000, title: 'settings.lockduration.30minutes' },
    { milliseconds: 3600000, title: 'settings.lockduration.1hour' },
    { milliseconds: 14400000, title: 'settings.lockduration.4hours' },
  ],
  aboutUrl: 'https://coinid.org',
  feedbackUrl: 'https://coinid.org/feedback',
  guidesUrl: 'https://coinid.org/guides',
  telegramUrl: 'https://t.me/joinchat/IARCoBAdhQOIEN_7u941Qg',
  twitterUrl: 'https://twitter.com/COINiDGroup',
  instagramUrl: 'https://www.instagram.com/coinidgroup',
  offlineGuideUrl: {
    ios: 'https://coinid.org/guides/preparing-ios-cold-wallet',
    android: 'https://coinid.org/guides/preparing-android-cold-wallet',
  },
  isTestnet : true,
};
