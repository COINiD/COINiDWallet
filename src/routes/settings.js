import { createStackNavigator } from 'react-navigation';
import SettingsRoute from '../components/SettingsRoute';

export const settingRoutes = {
  Home: {
    screen: SettingsRoute,
    swipeEnabled: false,
    title: 'Settings',
    isHome: true,
  },
  Passcode: {
    screen: SettingsRoute,
    title: 'Require unlocking',
  },
  OfflineTransport: {
    screen: SettingsRoute,
    title: 'Offline transport',
  },
  PreferredCurrency: {
    screen: SettingsRoute,
    title: 'Preferred currency',
  },
  SignMessage: {
    screen: SettingsRoute,
    title: 'Sign message',
  },
  About: {
    screen: SettingsRoute,
    title: 'About',
  },
  AccountList: {
    screen: SettingsRoute,
    title: 'Select an account',
  },
  AccountInformation: {
    screen: SettingsRoute,
    title: 'Account information',
  },
};

export const getSettingsNavigator = cardStyle => createStackNavigator(settingRoutes, {
  animationEnabled: true,
  headerMode: 'none',
  cardStyle,
});
