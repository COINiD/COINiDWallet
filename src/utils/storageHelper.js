/**
 * Helper for storage
 */

import { AsyncStorage } from 'react-native';

const getStorageItem = (namespace, key) => AsyncStorage.getItem(`@${namespace}:${key}`).then(JSON.parse);

const setStorageItem = (namespace, key, value) => {
  const val = JSON.stringify(value);
  return AsyncStorage.setItem(`@${namespace}:${key}`, val);
};

const removeStorageItem = (namespace, key) => AsyncStorage.removeItem(`@${namespace}:${key}`);

const reset = namespace => AsyncStorage.getAllKeys().then((keys) => {
  const filteredKeys = keys.filter(key => key.startsWith(`@${namespace}`));
  return AsyncStorage.multiRemove(filteredKeys);
});

export default namespace => ({
  get: key => getStorageItem(namespace, key),
  set: (key, value) => setStorageItem(namespace, key, value),
  remove: key => removeStorageItem(namespace, key),
  reset: () => reset(namespace),
});
