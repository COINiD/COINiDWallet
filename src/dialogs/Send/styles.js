import { StyleSheet } from 'react-native';
import { colors, fontWeight } from '../../config/styling';
import styleMerge from '../../utils/styleMerge';
import parentStyles from '../dialogStyles';

export default styleMerge(
  parentStyles('light'),
  StyleSheet.create({
    container: {
      paddingTop: 8,
    },
    deleteIcon: {
      color: '#FFFFFF',
      fontSize: 21,
    },
    deleteIconContainer: {
      // marginLeft: -10
    },
    currencyButton: {
      position: 'absolute',
      right: 0,
      bottom: 0,
      borderWidth: 1,
      borderColor: colors.purple,
      borderRadius: 8,
      paddingTop: 6,
      paddingBottom: 5,
      paddingHorizontal: 7,
      marginBottom: 8,
    },
    currencyButtonText: {
      color: '#617AF7',
      ...fontWeight.medium,
      letterSpacing: 0.1,
    },
  }),
);
