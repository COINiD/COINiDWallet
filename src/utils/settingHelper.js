'use strict';

/**
 * Helper for loading and saving settings.
 */

import { EventEmitter } from 'events';
import storageHelper from './storageHelper';
import Settings from '../config/settings';

class SettingHelper extends EventEmitter {
  constructor(coin) {
    super();

    this.setMaxListeners(50);

    this.storageNS = `${coin}:settings`;
    this.storage = storageHelper(this.storageNS);
    this.defaultSettings = {
      coldWalletMode: true,
      currencies: Settings.availableCurrencies,
      currency: Settings.currency,
      range: 0,
      usePasscode: true,
      lockAfterDuration: 60000,
      preferredColdTransport: '',
    };
    this.settings = { ...this.defaultSettings };
  }

  load = () => {
    const getDefaultIfEmpty = (key, data) =>
      data !== null ? data : this.defaultSettings[key];

    var p = Object.entries(this.defaultSettings).map(([key, value]) =>
      this.storage.get(key).then(data => getDefaultIfEmpty(key, data))
    );

    return Promise.all(p).then(
      ([
        coldWalletMode,
        currencies,
        currency,
        range,
        usePasscode,
        lockAfterDuration,
        preferredColdTransport,
      ]) => {
        this.settings = {
          coldWalletMode,
          currencies,
          currency,
          range,
          usePasscode,
          lockAfterDuration,
          preferredColdTransport,
        };

        this.emit('updated', this.settings);
      }
    );
  };

  update = (key, value) => {
    this.settings[key] = value;
    this.save();
    this.emit('updated', this.settings);
  };

  save = () => {
    var p = Object.entries(this.settings).map(([key, value]) => {
      return this.storage.set(key, value);
    });

    return Promise.all(p);
  };

  reset = () => {
    this.storage.reset();
  };

  getAll = () => {
    return this.settings;
  };
}

let settingHelpersCache = {}; // for local caching so we dont create several for same coin.

module.exports = function(coin) {
  if (settingHelpersCache[coin] === undefined) {
    settingHelpersCache[coin] = new SettingHelper(coin);
  }

  return settingHelpersCache[coin];
};
