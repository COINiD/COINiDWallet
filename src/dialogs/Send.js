import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  Alert, StyleSheet, Platform, TextInput, View, TouchableOpacity,
} from 'react-native';
import { Icon } from 'react-native-elements';
import Big from 'big.js';

import {
  Button, Text, AmountInput, FontScale,
} from '../components';
import { numFormat } from '../utils/numFormat';
import { fontSize, colors, fontWeight } from '../config/styling';
import { decodeQrRequest } from '../utils/addressHelper';

import WalletContext from '../contexts/WalletContext';
import { withExchangeRateContext } from '../contexts/ExchangeRateContext';

import styleMerge from '../utils/styleMerge';
import parentStyles from './styles/common';

const styles = styleMerge(
  parentStyles('light'),
  StyleSheet.create({
    container: {
      paddingTop: 8,
    },
    deleteIcon: {
      color: '#FFFFFF',
      fontSize: 21,
    },
    deleteIconContainer: {},
    currencyButton: {
      position: 'absolute',
      right: 0,
      bottom: 0,
      borderWidth: 1,
      borderColor: colors.purple,
      borderRadius: 8,
      paddingTop: 6,
      paddingBottom: 5,
      paddingHorizontal: 7,
      marginBottom: 8,
    },
    currencyButtonText: {
      color: '#617AF7',
      ...fontWeight.medium,
      letterSpacing: 0.1,
    },
  }),
);

class Send extends PureComponent {
  static contextType = WalletContext;

  static propTypes = {
    navigation: PropTypes.shape({}).isRequired,
    dialogRef: PropTypes.shape({}).isRequired,
    editItem: PropTypes.shape({}),
    fee: PropTypes.number,
    balance: PropTypes.number.isRequired,
    onAddToBatch: PropTypes.func.isRequired,
    onRemoveFromBatch: PropTypes.func.isRequired,
    exchangeRateContext: PropTypes.shape({}).isRequired,
  };

  static defaultProps = {
    editItem: {},
    fee: 0,
  };

  constructor(props, context) {
    super(props);

    const { navigation, editItem } = props;
    const { coinid } = context;
    const { ticker } = coinid;

    this.coinid = coinid;
    this.navigation = navigation;

    this.state = {
      exchangeRate: 0,
      ticker,
      inputInFiat: false,
      address: '',
      note: '',
      amount: '',
      editAddress: '',
      editAmount: 0,
    };

    if (editItem.address) {
      this.state = {
        ...this.state,
        address: editItem.address,
        note: editItem.note,
        amount: editItem.amount,
        editAddress: editItem.address,
        editAmount: editItem.amount,
      };
    }
  }

  _verify = () => {
    const { address } = this.state;
    let { amount } = this.state;

    amount = Number(amount);

    const errors = [];

    if (isNaN(amount)) {
      errors.push({
        type: 'amount',
        message: 'amount is not a number',
      });
    }

    if (amount === Number(0)) {
      errors.push({
        type: 'amount',
        message: 'amount cannot be zero',
      });
    }

    /*
    const { editAmount } = this.state;

    const { balance } = this.props;
    const availableBalance = balance + editAmount;

    if (amount > availableBalance) {
      errors.push({
        type: 'balance',
        message: 'not enough funds',
      });
    }
    */

    if (!this.coinid.validateAddress(address)) {
      errors.push({
        type: 'address',
        message: `address is not a valid ${this.coinid.coin} address`,
      });
    }

    if (errors.length) {
      throw errors;
    }

    return true;
  };

  _addToBatch = () => {
    const { onAddToBatch } = this.props;

    let { amount } = this.state;
    const { address, note, editAddress } = this.state;
    amount = Number(amount);

    onAddToBatch({ amount, address, note }, editAddress);
  };

  _removeFromBatch = () => {
    const { onRemoveFromBatch } = this.props;

    const { editAddress } = this.state;
    onRemoveFromBatch(editAddress);
  };

  _submit = () => {
    try {
      if (this._verify()) {
        this._addToBatch();
      }
    } catch (err) {
      Alert.alert(err[0].message);
    }
  };

  _startQr = () => {
    this.navigation.navigate('QRScan', { qrCodeResult: this._parseQRCodeResult });
  };

  _parseQRCodeResult = (qrResult) => {
    const {
      coinid: {
        network: { bech32, qrScheme },
      },
    } = this.context;

    const decoded = decodeQrRequest(qrResult, { qrScheme, bech32 });
    const { address, amount = '', note = '' } = decoded;

    if (address) {
      this.amountRef._updateAmount(amount);
      this.setState({ address, amount, note });
      return true;
    }

    return false;
  };

  _onChangeAmount = (amount) => {
    this.setState({ amount });
  };

  _toggleInputFiat = () => {
    const { inputInFiat } = this.state;
    const { exchangeRateContext } = this.props;
    const { exchangeRate } = exchangeRateContext;

    if (!exchangeRate) {
      return;
    }

    this.setState({ inputInFiat: !inputInFiat });
  };

  render() {
    const {
      editAddress, editAmount, address, note, ticker, amount,
    } = this.state;

    const {
      balance, dialogRef, fee, exchangeRateContext,
    } = this.props;
    const { exchangeRate, currency } = exchangeRateContext;

    let { inputInFiat } = this.state;

    if (!exchangeRate) {
      inputInFiat = false;
    }

    const availableBalance = Big(balance)
      .plus(editAmount)
      .minus(fee);

    const renderEditButton = () => (
      <View style={{ flexDirection: 'row' }}>
        <Button
          style={[
            styles.formButton,
            {
              backgroundColor: '#2A2937',
              maxWidth: 48,
              marginRight: 10,
            },
          ]}
          onPress={this._removeFromBatch}
        >
          <Icon
            name="delete"
            containerStyle={styles.deleteIconContainer}
            iconStyle={styles.deleteIcon}
          />
        </Button>
        <Button style={styles.formButton} onPress={this._submit}>
          Update
        </Button>
      </View>
    );

    const renderAvailableBalance = () => {
      const getAmount = () => {
        if (inputInFiat) {
          const fiatAmount = Big(availableBalance).times(exchangeRate);

          return `${numFormat(fiatAmount, currency)} ${currency}`;
        }

        return `${numFormat(availableBalance, ticker)} ${ticker}`;
      };

      return (
        <View style={{ flexDirection: 'row' }}>
          <Text style={styles.formInfo}>Available balance: </Text>
          <TouchableOpacity
            onPress={() => {
              if (availableBalance > 0) {
                this.amountRef._updateAmount(availableBalance);
              }
            }}
          >
            <Text style={[styles.formInfo, availableBalance < 0 ? styles.negativeBalance : null]}>
              {getAmount()}
            </Text>
          </TouchableOpacity>
        </View>
      );
    };

    const renderSendButton = () => (
      <Button style={styles.formButton} onPress={this._submit}>
        Add transaction
      </Button>
    );

    return (
      <View
        style={styles.container}
        ref={(c) => {
          this.refCont = c;
        }}
        onLayout={() => {}}
      >
        <View
          style={styles.modalContent}
          onLayout={(e) => {
            this.refContHeight = e.nativeEvent.layout.height;
          }}
        >
          <View
            style={styles.formItem}
            onFocus={() => {
              dialogRef._setKeyboardOffset(this.refToBottom - this.refContHeight + 8);
            }}
            onLayout={(e) => {
              this.refToBottom = e.nativeEvent.layout.y + e.nativeEvent.layout.height;
            }}
          >
            <Text style={styles.formLabel}>To</Text>
            <View style={styles.formItemRow}>
              <FontScale
                style={{
                  flex: 1,
                  flexDirection: 'column',
                  alignItems: 'flex-start',
                }}
                fontSizeMax={fontSize.base}
                fontSizeMin={6}
                text={address.trim()}
                widthScale={0.9}
                ref={(c) => {
                  this.toScaleRef = c;
                }}
              >
                {({ fontSize: scaledFontSize }) => (
                  <TextInput
                    keyboardType={Platform.OS === 'ios' ? 'default' : 'visible-password'}
                    style={[
                      styles.formItemInput,
                      { flex: 0, width: '100%', fontSize: scaledFontSize },
                    ]}
                    value={address}
                    autoCorrect={false}
                    spellCheck={false}
                    textContentType="none"
                    onChangeText={(changedAddress) => {
                      this.setState({ address: changedAddress.trim() });
                    }}
                    onBlur={() => {
                      const orgAddress = address;

                      this.setState({ address: `${orgAddress} ` }, () => {
                        setTimeout(() => {
                          this.setState({ address: `${orgAddress}` });
                        }, 100);
                      });
                    }}
                    ref={(c) => {
                      this.toRef = c;
                    }}
                    returnKeyType="done"
                    onSubmitEditing={() => this.amountRef.focus()}
                    underlineColorAndroid="transparent"
                    allowFontScaling={false}
                  />
                )}
              </FontScale>
              <View style={styles.formItemIcons}>
                <Icon
                  iconStyle={styles.formItemIcon}
                  type="material-community"
                  name="qrcode"
                  onPress={this._startQr}
                  hitSlop={{
                    top: 20,
                    bottom: 20,
                    left: 0,
                    right: 20,
                  }}
                />
              </View>
            </View>
          </View>

          <View
            style={styles.formItem}
            ref={(c) => {
              this.amountContRef = c;
            }}
            onFocus={() => {
              dialogRef._setKeyboardOffset(this.refAmountBottom - this.refContHeight + 8);
            }}
            onLayout={(e) => {
              this.refAmountBottom = e.nativeEvent.layout.y + e.nativeEvent.layout.height;
            }}
          >
            <View>
              <Text style={styles.formLabel}>Amount</Text>
              <View style={styles.formItemRow}>
                <AmountInput
                  ref={(c) => {
                    this.amountRef = c;
                  }}
                  style={[styles.formItemInput, { paddingRight: 60 }]}
                  onChangeAmount={this._onChangeAmount}
                  onSubmitEditing={() => this.noteRef.focus()}
                  exchangeRate={exchangeRate}
                  inputInFiat={inputInFiat}
                  amount={amount}
                  exchangeTo={currency}
                  exchangeFrom={ticker}
                />
              </View>
              <TouchableOpacity style={styles.currencyButton} onPress={this._toggleInputFiat}>
                <Text style={styles.currencyButtonText}>{inputInFiat ? currency : ticker}</Text>
              </TouchableOpacity>
            </View>
            {renderAvailableBalance()}
          </View>

          <View
            style={styles.formItem}
            onFocus={() => {
              dialogRef._setKeyboardOffset(this.refNoteBottom - this.refContHeight + 8);
            }}
            onLayout={(e) => {
              this.refNoteBottom = e.nativeEvent.layout.y + e.nativeEvent.layout.height;
            }}
          >
            <Text style={styles.formLabel}>Note</Text>
            <View style={styles.formItemRow}>
              <TextInput
                keyboardType={Platform.OS === 'ios' ? 'default' : 'visible-password'}
                autoCorrect={false}
                spellCheck={false}
                textContentType="none"
                style={styles.formItemInput}
                value={note}
                maxLength={26}
                onChangeText={newNote => this.setState({ note: newNote })}
                ref={(c) => {
                  this.noteRef = c;
                }}
                returnKeyType="done"
                onSubmitEditing={() => this.noteRef.blur()}
                underlineColorAndroid="transparent"
              />
            </View>
          </View>

          {editAddress ? renderEditButton() : renderSendButton()}
        </View>
      </View>
    );
  }
}

export default withExchangeRateContext()(Send);
