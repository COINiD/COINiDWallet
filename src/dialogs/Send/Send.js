

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  Platform,
  Clipboard,
  Keyboard,
  TouchableOpacity,
  TextInput,
  View,
  KeyboardAvoidingView,
} from 'react-native';
import { Icon } from 'react-native-elements';
import { Button, Text, DetailsModal } from '../../components';
import styles from './styles';
import { numFormat } from '../../utils/numFormat';

export default class Send extends PureComponent {
  constructor(props) {
    super(props);

    const { navigation } = this.props;

    this.navigate = navigation;

    this.state = {
      address: undefined,
      amount: '',
      note: undefined,
      editAddress: '',
      editAmount: 0,
    };
  }

  componentDidMount() {
    const { ticker } = this.context.coinid;
    this.setState({ ticker });
  }

  getChildContext() {
    return {
      theme: this.props.theme ? this.props.theme : this.context.theme,
    };
  }

  _verify = () => {
    let { amount, address, editAmount } = this.state;
    const { balance } = this.props;
    const availableBalance = balance + editAmount;

    amount = Number(amount);

    const errors = [];

    if (isNaN(amount)) {
      errors.push({
        type: 'amount',
        message: 'amount is not a number',
      });
    }

    if (amount === Number(0)) {
      errors.push({
        type: 'amount',
        message: 'amount cannot be zero',
      });
    }

/*
    if (amount > availableBalance) {
      errors.push({
        type: 'balance',
        message: 'not enough funds',
      });
    }
*/

    if (!this.context.coinid.validateAddress(address)) {
      errors.push({
        type: 'address',
        message:
          `address is not a valid ${this.context.coinid.coin} address`,
      });
    }

    if (errors.length) {
      throw errors;
    }

    return true;
  };

  _addToBatch = () => {
    let {
      amount, address, note, editAddress,
    } = this.state;
    amount = Number(amount);
    this.props.onAddToBatch({ amount, address, note }, editAddress);
  };

  _removeFromBatch = () => {
    const { editAddress } = this.state;
    this.props.onRemoveFromBatch(editAddress);
  };

  _submit = () => {
    try {
      if (this._verify()) {
        this._addToBatch();
        // this._close();
      }
    } catch (err) {
      alert(err[0].message);
    }
  };

  _startQr = () => {
    this.navigate('QRScan', { qrCodeResult: this._parseQRCodeResult });
  };

  _parseQRCodeResult = (qrResult) => {
    if (qrResult.indexOf(':') > 0) {
      var address = qrResult.match(/[1-9A-HJ-NP-Za-km-z]{26,36}/g);
      if (address) {
        address = address[0];
        this.setState({ address });

        let uriAmount = qrResult.match(/=[0-9\.]+/g);

        if (uriAmount != null) {
          uriAmount = uriAmount[0].replace('=', '');
        }

        if (uriAmount) {
          this.setState({ amount: uriAmount });
        }
      }
    } else {
      var address = qrResult.match(/[1-9A-HJ-NP-Za-km-z]{26,36}/g);
      address = address[0];
      this.setState({ address });
    }
  };

  _open = (item) => {
    // if item is set then we enter edit mode
    if (item !== undefined) {
      this.setState({
        ...item,
        editAddress: item.address,
        editAmount: item.amount,
      });
    } else {
      this.setState({
        address: '',
        note: '',
        amount: '',
        editAddress: '',
        editAmount: 0,
      });
    }
    this.elModal._open();
  };

  _close = () => {
    this.elModal._close();
  };

  _isValidInput = (text) => {
    const regex = /^([0-9]\d*(\.\d{0,8})?)?$/;
    return regex.test(text);
  };

  _onChangeAmountText = (inputAmount) => {
    const amount = inputAmount
      .replace(',', '.')
      .replace(/^\./, '0.')
      .replace(/^[0]{2,}/, '0')
      .replace(/([.]\d{8,8})(.*$)/, '$1');

    if (!this._isValidInput(amount)) {
      return false;
    }

    this.setState({ amount });
  };

  _getAmountMaxLength = (amount) => {
    const maxLength1 = `${amount}`.replace(/\.[0-9]*$/, '').length + 8 + 1;
    const maxLength2 = `${amount}`.length + 1;

    return maxLength1 < maxLength2 ? maxLength1 : maxLength2;
  };

  render() {
    const {
      editAddress,
      editAmount,
      amount,
      address,
      note,
      ticker,
    } = this.state;
    const { onOpened, onClosed, balance } = this.props;

    const availableBalance = balance + editAmount;

    const renderEditButton = () => (
      <View style={{ flexDirection: 'row' }}>
        <Button
          style={[
            styles.formButton,
            {
              backgroundColor: '#2A2937',
              maxWidth: 48,
              marginRight: 10,
            },
          ]}
          onPress={this._removeFromBatch}
        >
          <Icon
            name="delete"
            containerStyle={styles.deleteIconContainer}
            iconStyle={styles.deleteIcon}
          />
        </Button>
        <Button style={styles.formButton} onPress={this._submit}>
            Update
        </Button>
      </View>
    );

    const renderSendButton = () => (
      <Button style={styles.formButton} onPress={this._submit}>
          Add transaction
      </Button>
    );

    return (
      <DetailsModal
        ref={(c) => { this.elModal = c; }}
        title={editAddress ? 'Edit Transaction' : 'Send'}
        verticalPosition="flex-end"
        onOpened={onOpened}
        onClosed={onClosed}
        avoidKeyboard
        avoidKeyboardOffset={40}
      >
        <View
          style={styles.container}
          ref={(c) => {
            this.refCont = c;
          }}
          onLayout={() => {}}
        >
          <View
            style={styles.modalContent}
            onLayout={(e) => { this.refContHeight = e.nativeEvent.layout.height; }}
          >
            <View
              style={styles.formItem}
              onLayout={c => this.toContPos = c}
              ref={c => this.toContRef = c}
              onFocus={(e) => { this.elModal._setKeyboardOffset(this.refToBottom - this.refContHeight + 8); }}
              onLayout={(e) => { this.refToBottom = e.nativeEvent.layout.y + e.nativeEvent.layout.height; }}
            >
              <Text style={styles.formLabel}>To</Text>
              <View style={styles.formItemRow}>
                <TextInput
                  keyboardType={Platform.OS === 'ios' ? 'default' : 'visible-password'}
                  style={styles.formItemInput}
                  value={address}
                  autoCorrect={false}
                  spellCheck={false}
                  textContentType={false}
                  onChangeText={address => this.setState({ address: address.trim() })}
                  ref={(c) => { this.toRef = c; }}
                  returnKeyType="done"
                  onSubmitEditing={() => this.amountRef.focus()}
                  underlineColorAndroid="transparent"
                />
                <View style={styles.formItemIcons}>
                  <Icon
                    iconStyle={styles.formItemIcon}
                    type="material-community"
                    name="qrcode"
                    onPress={this._startQr}
                    hitSlop={{
                      top: 20,
                      bottom: 20,
                      left: 20,
                      right: 20,
                    }}
                  />
                </View>
              </View>
            </View>

            <View
              style={styles.formItem}
              ref={c => this.amountContRef = c}
              onFocus={(e) => { this.elModal._setKeyboardOffset(this.refAmountBottom - this.refContHeight + 8); }}
              onLayout={(e) => { this.refAmountBottom = e.nativeEvent.layout.y + e.nativeEvent.layout.height; }}
            >
              <Text style={styles.formLabel}>Amount</Text>
              <View style={styles.formItemRow}>
                <TextInput
                  ref={c => (this.amountRef = c)}
                  keyboardType="numeric"
                  style={styles.formItemInput}
                  value={`${amount}`}
                  onChangeText={this._onChangeAmountText}
                  returnKeyType="done"
                  onSubmitEditing={() => this.noteRef.focus()}
                  underlineColorAndroid="transparent"
                  maxLength={this._getAmountMaxLength(amount)}
                />
              </View>
              <View style={{ flexDirection: 'row' }}>
                <Text style={styles.formInfo}>{`Available balance: `}</Text>
                <Text style={[styles.formInfo, (availableBalance < 0 ? styles.negativeBalance : null)]}>{`${numFormat(availableBalance, ticker)} ${ticker}`}</Text>
              </View>
            </View>

            <View
              style={styles.formItem}
              onFocus={(e) => { this.elModal._setKeyboardOffset(this.refNoteBottom - this.refContHeight + 8); }}
              onLayout={(e) => { this.refNoteBottom = e.nativeEvent.layout.y + e.nativeEvent.layout.height; }}
            >
              <Text style={styles.formLabel}>Note</Text>
              <View style={styles.formItemRow}>
                <TextInput
                  keyboardType={Platform.OS === 'ios' ? 'default' : 'visible-password'}
                  autoCorrect={false}
                  spellCheck={false}
                  textContentType={false}
                  style={styles.formItemInput}
                  value={note}
                  maxLength={26}
                  onChangeText={note => this.setState({ note })}
                  ref={c => (this.noteRef = c)}
                  returnKeyType="done"
                  onSubmitEditing={() => this.noteRef.blur()}
                  underlineColorAndroid="transparent"
                />
              </View>
            </View>

            {editAddress ? renderEditButton() : renderSendButton()}
          </View>
        </View>
      </DetailsModal>
    );
  }
}

Send.contextTypes = {
  coinid: PropTypes.object,
  type: PropTypes.string,
  theme: PropTypes.string,
  settingHelper: PropTypes.object,
};

Send.childContextTypes = {
  theme: PropTypes.string,
};

Send.propTypes = {
  theme: PropTypes.string,
};

Send.defaultProps = {
  theme: 'light',
};
