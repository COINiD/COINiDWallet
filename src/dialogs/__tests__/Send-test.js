import 'react-native';
import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import PropTypes from 'prop-types';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });
import Send from '../Send';

import COINiDPublic from '../../libs/coinid-public';
import settings from '../../config/settings';
import storageHelper from '../../utils/storageHelper';
import settingHelper from '../../utils/settingHelper';

const context = {
  coinid: COINiDPublic(
    settings.coin,
    storageHelper(`${settings.coin}-hot`),
    `${settings.coin}-hot`,
  ),
  settingHelper: settingHelper(settings.coin),
}

describe('decodeQrRequest', () => {

  const getWrapper = (props={balance:10})Â => {
    const wrapper = shallow(
      <Send {...props} />,
      { context },
    );

    const instance = wrapper.instance();
    instance.amountRef = { _updateAmount: jest.fn() };

    return { wrapper, instance };
  }

  it('parses a qr code result', () => {
    const { wrapper, instance } = getWrapper();

    const isParsed = instance._parseQRCodeResult('bitcoin:mhNR7zLchYfX5GtFZ1opW8FeAG63wVjP9D?amount=10&message=Hello');
    expect(isParsed).toEqual(true);

    expect(wrapper.state()).toMatchSnapshot();

    expect(instance._verify).not.toThrow();
    wrapper.setState({ amount: "0", address: "mhNR7zLchYfX5GtFZ1opW8FeAG63wVjP9D" });
    expect(instance._verify).toThrow();
    wrapper.setState({ amount: "10", address: "123" });
    expect(instance._verify).toThrow();
    wrapper.setState({ amount: "10", address: "mhNR7zLchYfX5GtFZ1opW8FeAG63wVjP9D" });
    expect(instance._verify).not.toThrow();
    wrapper.setState({ amount: "100", address: "mhNR7zLchYfX5GtFZ1opW8FeAG63wVjP9D" });
    expect(instance._verify).not.toThrow();
  });

  it('not parse qr result', () => {
    const { wrapper, instance } = getWrapper();

    const isParsed = instance._parseQRCodeResult('wierdstring');
    expect(isParsed).toEqual(false);

    expect(wrapper.state()).toMatchSnapshot();
    expect(instance._verify).toThrow();
  });
});
