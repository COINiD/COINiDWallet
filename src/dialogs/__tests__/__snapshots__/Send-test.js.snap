// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decodeQrRequest not parse qr result 1`] = `
Object {
  "address": "",
  "amount": "",
  "currency": "USD",
  "editAddress": "",
  "editAmount": 0,
  "exchangeRate": 0,
  "inputInFiat": false,
  "note": undefined,
  "ticker": "tBTC",
}
`;

exports[`decodeQrRequest parses a qr code result 1`] = `
Object {
  "address": "mhNR7zLchYfX5GtFZ1opW8FeAG63wVjP9D",
  "amount": "10",
  "currency": "USD",
  "editAddress": "",
  "editAmount": 0,
  "exchangeRate": 0,
  "inputInFiat": false,
  "note": "Hello",
  "ticker": "tBTC",
}
`;
